feat: Staggered Grid Layout 컴포넌트 구현

HJunng opened the pull request:
## 📝 변경 사항
핀터레스트 스타일의 Staggered Grid Layout 컴포저블을 구현하여 장소 이미지를 표시할 수 있도록 했습니다.

## 🎯 작업 내용
- `MemoripStaggeredGrid` 컴포저블 구현
  - 2열 구조의 엇갈린 그리드 레이아웃 구현
  - 이미지 높이에 따라 자동으로 배치되는 반응형 레이아웃
- Coil3 라이브러리를 사용한 비동기 이미지 로딩 구현
- 디자인 시스템 구성 요소 추가
  - 커스텀 Gray 색상 추가
  - 앱 전반에서 사용될 Dimen 정의
  - 장소 이미지 관련 문자열 리소스 추가
- 인터넷 권한 추가 (이미지 로딩을 위해)
- 임시로 랜덤 이미지 URL을 생성하는 유틸 함수 `generateRandomImageUrls` 추가
- 코드 정리 및 리팩토링
  - `ImageItem` 데이터 클래스 위치 변경 및 주석 추가
  - @Preview 코드 제거

## 🔗 관련 이슈
Closes #5

## 📸 스크린샷 (선택사항)
<img width="50%" alt="image" src="https://github.com/user-attachments/assets/963d9f62-6f6f-4479-80ed-bb75b03f2cd3" />

- 사진 크기와 상관없이 비율을 width 최소 160dp 이상이 되도록 배수 처리를 해서
- 320dp-479dp까지는 2줄로, 480-639dp까지는 3줄, ... 이런식으로 보여지게 함.
- 160dp로 둔 이유는 저희 앱 minSDK 24이상의 기기 중에서 가장 작은 너비 dpi가 320인 것을 확인함.
- 해당 기기부터 2줄로 보여지도록 지원하기 위해 160dp로 할 예정.

<img width="398" height="888" alt="image" src="https://github.com/user-attachments/assets/0130968d-35e4-4293-be3b-eb917262f280" />

- 기존의 구현에서는 주어진 width 비율에 맞춰서 전체 height를 채워야 했어서 preview를 그릴 수 없었는데 
- 사용자 지정으로 height를 설정할 수 있는 옵션을 추가해서, 파라미터로 height를 추가하면 고정된 크기로 보여줄 수 있도록 추가하여 
- 기능도 범용적으로 만들고, preview 도 추가하였습니다.



## 💬 리뷰어에게
- StaggeredGrid는 이미지 원본 크기에 따라 동적으로 레이아웃이 결정되는데, 
Preview 환경에서는 네트워크 이미지 로드가 불가하여 레이아웃 확인이 어려워 Preview를 제거했습니다.
이 컴포넌트는 Preview 없이 진행해도 괜찮을까요? 아니면 Preview를 위한 더 좋은 방법이 있을까요?
- 추후 실제 데이터 연결 시 `generateRandomImageUrls` 함수를 제거하고 실제 API 데이터로 교체할 예정입니다.

HJunng committed:
feat: Coil 라이브러리 의존성 추가

HJunng committed:
feat: 인터넷 권한 추가

HJunng committed:
feat: 커스텀 색상 Gray 추가

HJunng committed:
feat: 앱 전반에서 사용될 Dimen 정의

HJunng committed:
feat: 장소 이미지 문자열 리소스 추가

HJunng committed:
feat: StaggeredGrid 컴포저블 구현

- 핀터레스트 화면 기준으로 StaggeredGrid 사용
- 이미지 로딩을 위해 Coil3 사용
- 임시로 사용할 랜덤 이미지 URL 생성 함수 `generateRandomImageUrls`를 추가해서 활용

HJunng committed:
fix: MemoripStaggeredGrid 코드 정리

- `ImageItem` 데이터 클래스의 위치를 변경하고, 주석 추가
- @Preview 코드 제거

uselessnaming commented on client/presentation/src/main/java/com/andone/memorip/presentation/component/MemoripStaggeredGrid.kt:
```diff
@@ -0,0 +1,98 @@
+
+@Composable
+fun MemoripStaggeredGrid(
+    modifier: Modifier = Modifier,
```
이 부분 paramter 1개면 1줄에 작성하기로 했습니다!

uselessnaming reviewed (approved):
preview는 있으면 좋을 것 같습니다! 원택님이 하신 것처럼 preview해주시면 없는 이미지로 대충 볼 수 있을 것 같습니다!
그리고 혹시 가로로도 다르면 어떻게 되는지도 보여주면 좋을 것 같아요!

HJunng committed:
fix: StaggeredGrid 컴포저블 코드 정리

- `StaggeredGridDimens` 오브젝트를 추가하여 현재 화면 관련 Dimen 값 관리
- 랜덤 이미지의 최소 높이를 50으로 수정
- `StaggeredImageItem`에서 Modifier 적용 순서 변경

hoyadong1 commented on client/presentation/src/main/java/com/andone/memorip/presentation/theme/Dimen.kt:
디멘 좋네요 잘 쓰겠습니다

GwonDongHyeon21 committed:
feat: Dimen 추가

HJunng committed:
feat: MemoripStaggeredGrid에 공통 Dimen 반영

HJunng committed:
feat: StaggeredGrid 관련 문자열 리소스 이름 변경

HJunng committed:
feat: MemoripStaggeredGrid 미리보기 추가

- `StaggeredImageItem`에 고정 높이를 설정할 수 있는 `fixedHeight` 파라미터를 추가했습니다.
- `LazyVerticalStaggeredGrid`를 사용한 미리보기(`@Preview`)를 추가하여 다양한 높이의 이미지가 표시되는 것을 확인할 수 있습니다.
- 미리보기에서 사용할 더미 데이터 높이를 생성하는 로직을 추가했습니다.

HJunng commented:
> preview는 있으면 좋을 것 같습니다! 원택님이 하신 것처럼 preview해주시면 없는 이미지로 대충 볼 수 있을 것 같습니다!
그리고 혹시 가로로도 다르면 어떻게 되는지도 보여주면 좋을 것 같아요!

이 부분 수정해서 위의 PR 내용 수정해두었습니다~

GwonDongHyeon21 commented on client/presentation/src/main/res/values/strings.xml:
```diff
@@ -1,3 +1,6 @@
     <string name="app_name">Memorip</string>
+
+    <!-- StaggeredGrid-->
+    <string name="staggered_grid_image_contentDescription">장소 이미지</string>
```
contentDescription -> content_description으로 작성해야하지 않을까요?

HJunng replied:
> 화면이름_목적_contentDescription / 컴포넌트이름_목적_contentDescription

노션에 이렇게 쓰여져 있어가지고 이거 그대로 가져다가 썼는데
content_description으로 하는 건가욤?

GwonDongHyeon21 replied:
어머 그렇군요. content_description 이게 더 맞다고 생각합니다!
노션도 수정해놓겠습니다.

GwonDongHyeon21 commented on client/presentation/src/main/java/com/andone/memorip/presentation/component/MemoripStaggeredGrid.kt:
```diff
@@ -0,0 +1,134 @@
+            .background(MemoripTheme.colors.gray),
+        contentAlignment = Alignment.Center
+    ) {
+        AsyncImage(
+            model = imageUrl,
+            contentDescription = contentDescription,
+            contentScale = ContentScale.FillWidth,
+            placeholder = ColorPainter(MemoripTheme.colors.gray),
+            error = ColorPainter(MemoripTheme.colors.gray),
+            modifier = Modifier.fillMaxWidth()
+        )
```
parameter 선언부와 호출부 순서를 일치시키면 좋을 것 같습니다!

HJunng replied:
아 이거 AsyncImage를 호출하는 MemoripStaggeredGrid 에 있는 파라미터랑 호출 순서를 일치시키라는 거였나요?? 
AsyncImage 내부에 있는 파라미터랑 맞추는게 아니었나요??

GwonDongHyeon21 replied:
AsyncImage 내부 파라미터 순서가 지금 호출된 순서랑 다르지 않나요?

HJunng replied:
아..ㅋㅋ 제가 그냥 coil 쓰고 있었습니다.. 그거에 맞춰서 파라미터 선언해서 쓰고 있었네요 
수정 했습니다!!! 감삼다

GwonDongHyeon21 commented on client/presentation/src/main/java/com/andone/memorip/presentation/component/MemoripStaggeredGrid.kt:
```diff
@@ -0,0 +1,134 @@
+            items(images) { image ->
+                StaggeredImageItem(
+                    imageUrl = image.url
+                )
```
parameter가 하나면 한줄로 작성하기로 했습니다!

HJunng replied:
수정했습니다!

GwonDongHyeon21 commented on client/presentation/src/main/java/com/andone/memorip/presentation/component/MemoripStaggeredGrid.kt:
```diff
@@ -0,0 +1,134 @@
+data class ImageItem(
+    val id: Int,
+    val url: String
+)
```
model 패키지 만들어서 분리해두면 좋을 것 같습니다.

GwonDongHyeon21 reviewed (approved):
고생하셨습니다. 랜덤이미지로 테스트 해본거 좋네요!

GwonDongHyeon21 commented on client/app/src/main/AndroidManifest.xml:
```diff
@@ -1,6 +1,7 @@
 <?xml version="1.0" encoding="utf-8"?>
 <manifest xmlns:android="http://schemas.android.com/apk/res/android">
 
+    <uses-permission android:name="android.permission.INTERNET" />
```
혹시 이부분 presentation 모듈의 AndroidManifest에 추가해도 정상동작하나요? 
정상동작하면 옮기는게 좋지 않을까 생각이 듭니다.

HJunng replied:
오 한번 확인해보겠습니다!!

HJunng replied:
presentation 모듈에 넣어서 정상 동작 하는 거 확인 후, 수정해두었습니다!

HJunng committed:
Merge branch 'dev' into feature/5-component-staggered-grid

# Conflicts:
#	client/gradle/libs.versions.toml
#	client/presentation/build.gradle.kts
#	client/presentation/src/main/java/com/andone/memorip/presentation/theme/Dimen.kt
#	client/presentation/src/main/res/values/strings.xml

HJunng committed:
fix: MemoripStaggeredGrid 코드 정리

- `ImageItem` 데이터 클래스를 `model` 패키지로 이동
- `AsyncImage`의 `contentScale` 파라미터 순서 변경
- coil 라이브러리 의존성 변경

HJunng committed:
feat: 인터넷 권한 선언 위치 변경

- app 모듈에서 presentation 모듈로 인터넷 권한 선언 위치를 이동
- 인터넷을 사용하는 주체는 presentation 모듈이므로 해당 모듈에 권한을 선언

HJunng committed:
fix: contentDescription 문자열 리소스 이름 변경

- 리소스 이름의 일관성을 위해 `contentDescription`을 `content_description`으로 변경

HJunng committed:
fix: contentDescription 문자열 리소스 이름 변경

- 리소스 이름의 일관성을 위해 `contentDescription`을 `content_description`으로 변경

GwonDongHyeon21 committed:
Merge branch 'dev' into feature/5-component-staggered-grid
